External/FEXCore/Source/Interface/Context/Context.h

```diff
@@ -269,9 +269,23 @@ namespace FEXCore::Context {
 
     uint8_t GetGPRSize() const { return Config.Is64BitMode ? 8 : 4; }
 
+    std::shared_mutex MemoryEntryMutex{};
```

Comment:
I don't think this should be moved to its own class just yet - it is unclear from where we'll need access to the data before everything gets implemented.

`AddMemoryRange` would be misleading as well, as maps might replace previous mappings.

There's also some information that is tracked in `AOTIRCaptureCache` now, which also needs to be removed from there and merged with the memory mapping tracking logic.

Overall, I'm a bit wary splitting things aggressively to classes, as we often need access to the data structures and locks in ways and places that don't nicely map with the class splitting.

Case in point the LookupCache class, where we end up accessing several internal structures and locks from Core - so there is little benefit from that split. And there's the awkward part of the LocalIRCache that resides outside the LookupCache.

Unfortunately, it also looks like the code will need to implement several "linux-isms", which makes me wonder if it is best to delegate more of this logic to the syscall side.

Reply:
> AddMemoryRange would be misleading as well, as maps might replace previous mappings.

It's easy to come up with a different verb. The key suggestion is to talk about memory ranges rather than maps, since the former are the main subject of this function.

> I don't think this should be moved to its own class just yet - it is unclear from where we'll need access to the data before everything gets implemented.

> There's also some information that is tracked in AOTIRCaptureCache now, which also needs to be removed from there and merged with the memory mapping tracking logic.

If it's unclear who needs access to this data, and if there's related information spread over two different modules, then it sounds like the design needs more thought?

I don't know which state in AOTIRCaptureCache you're referring to (seems like it wasn't introduced in this PR?), but as far as the first point goes it would be better to move the new Context state to a class now and adjust the design accordingly if we see a concrete need for other modules to access this data.

> Overall, I'm a bit wary splitting things aggressively to classes, as we often need access to the data structures and locks in ways and places that don't nicely map with the class splitting.

> Case in point the LookupCache class, where we end up accessing several internal structures and locks from Core - so there is little benefit from that split. And there's the awkward part of the LocalIRCache that resides outside the LookupCache.

To me, this indicates the code structure needed more thought (going beyond just wrapping data into classes) rather than a reason to throw everything together. The latter does make it easier to write code, but that comes at the expense of making it inaccessible for new readers and making it difficult to safely extend without regressing.

Reply:
> It's easy to come up with a different verb. The key suggestion is to talk about memory ranges rather than maps, since the former are the main subject of this function.

The reason I stuck with the map naming is because the code is "tracking the memory map calls". Anyway, I don't have a strong preference on any name, and those functions will be replaced with other ones in the next iterations (to handle shm, mremap and shared mappings) so I don't think this naming really matters.

> I don't know which state in AOTIRCaptureCache you're referring to (seems like it wasn't introduced in this PR?), but as far as the first point goes it would be better to move the new Context state to a class now and adjust the design accordingly if we see a concrete need for other modules to access this data.

Yes, it was added in the AOT PRs, and then got refactored to its own class by Ryan. It's the `AddNamedRegion`/`RemoveNamedRegion` api and the `AOTIRCaptureCache::AddrToFile`/`AOTIRCaptureCache::AddrToFileEntry` data structures. It tracks partially the same information, but doesn't support range splitting or overlapping ranges, and also tracks the filename.

Again, I'd rather work with it in the context for now, until this further stabilizes. I'd rather not have the extra files, abstraction work, heap allocations, pointer dereferences, etc, before we have a better understanding where this needs to live, and where we need access from.

> To me, this indicates the code structure needed more thought (going beyond just wrapping data into classes) rather than a reason to throw everything together. The latter does make it easier to write code, but that comes at the expense of making it inaccessible for new readers and making it difficult to safely extend without regressing.

Agreed. Largely, the design has shifted many times and at times the code hasn't been refactored/cleaned up. There are many places in FEX where things need to be re-thought now that we are much further, though I don't think this PR is where this needs to happen.

In general, I'm a bit wary of OOP/Very abstracted code in emulators as things often don't map nicely, and having to through several layers of leaky abstractions can make the code harder to understand, and slower to execute.

Reply:
Big Design (tm) talk aside, we're luckily talking about a fairly minor encapsulation change here. I also don't mind leaving the bits in AOTIRCaptureCache unchanged for now, since they don't really interact with the state added in this PR anyway.

> The reason I stuck with the map naming is because the code is "tracking the memory map calls". Anyway, I don't have a strong preference on any name, and those functions will be replaced with other ones in the next iterations (to handle shm, mremap and shared mappings) so I don't think this naming really matters.

Respectfully, "I'll replace this anyway" strikes me as an odd thing to say about code you're asking to be reviewed.

> I'd rather not have the extra files, abstraction work, heap allocations, pointer dereferences, etc, before we have a better understanding where this needs to live, and where we need access from.

There are no extra files/abstraction work/heap allocations/pointer derefs needed for this proposal:
```cpp
struct PickAName {
    std::shared_mutex MemoryEntryMutex{};
    // ...
};

struct Context { 
    PickAName name_picker;
};
```

Reply:
@neobrain How about the structure in https://github.com/FEX-Emu/FEX/blob/skmp/smc-memtrack-2/Source/Tests/LinuxSyscalls/Syscalls.cpp#L695 && https://github.com/FEX-Emu/FEX/blob/skmp/smc-memtrack-2/Source/Tests/LinuxSyscalls/Syscalls.h#L169 ?

I did also try to keep them private, but that wasn't workable as I needed external access for some callback.

Reply:
> How about the structure in https://github.com/FEX-Emu/FEX/blob/skmp/smc-memtrack-2/Source/Tests/LinuxSyscalls/Syscalls.cpp#L695 && https://github.com/FEX-Emu/FEX/blob/skmp/smc-memtrack-2/Source/Tests/LinuxSyscalls/Syscalls.h#L169 ?

I like that SyscallHandler is not nearly as blown up as Context is.

I would still opt for grouping the various member functions (excluding `MarkGuestExecutableRange`) and VMATracking into a dedicated object that is a member of SyscallHandler. Alternatively, SyscallHandler could inherit from the new object to reduce syntax overhead at call-sites. There wouldn't be any difference as far as performance/call-site syntax goes then, but it would just clean up the definition.

> I did also try to keep them private, but that wasn't workable as I needed external access for some callback.

Hard to think about this without seeing the callback you're referring to :)